detox/src/devices/android/Emulator.js

```diff
@@ -24,45 +24,55 @@ class Emulator {
   }
 
   async boot(emulatorName) {
-    const headless = argparse.getArgValue('headless') ? '-no-window' : '';
-    const cmd = `-verbose -gpu host -no-audio ${headless} @${emulatorName}`;
-    log.verbose(this.emulatorBin, cmd);
+    const emulatorArgs = _.compact([
+      '-verbose',
+      '-gpu', 'auto',
+      '-no-audio',
+      argparse.getArgValue('headless') ? '-no-window' : '',
+      `@${emulatorName}`
+    ]);
+
+    let childProcessOutput;
     const tempLog = `./${emulatorName}.log`;
     const stdout = fs.openSync(tempLog, 'a');
     const stderr = fs.openSync(tempLog, 'a');
-    const tail = new Tail(tempLog);
-    const promise = spawn(this.emulatorBin, _.split(cmd, ' '), {detached: true, stdio: ['ignore', stdout, stderr]});
-
-    const childProcess = promise.childProcess;
-    childProcess.unref();
-
-    tail.on("line", function(data) {
-      if (data.includes('Adb connected, start proxing data')) {
-        detach();
-      }
-      if (data.includes(`There's another emulator instance running with the current AVD`)) {
-        detach();
+    const tail = new Tail(tempLog).on("line", (line) => {
+      if (line.includes('Adb connected, start proxing data')) {
+        childProcessPromise._cpResolve();
       }
     });
 
-    tail.on("error", function(error) {
-      detach();
-      log.verbose('Emulator stderr: ', error);
-    });
+    function detach() {
+      if (childProcessOutput) {
+        return;
+      }
 
-    promise.catch(function(err) {
-      log.error('Emulator ERROR: ', err);
-    });
+      childProcessOutput = fs.readFileSync(tempLog, 'utf8');
 
-    function detach() {
       tail.unwatch();
       fs.closeSync(stdout);
       fs.closeSync(stderr);
-      fs.unlink(tempLog, () => {});
-      promise._cpResolve();
+      fs.unlink(tempLog, _.noop);
     }
 
-    return promise;
+    log.verbose(this.emulatorBin, ...emulatorArgs);
+    const childProcessPromise = spawn(this.emulatorBin, emulatorArgs, { detached: true, stdio: ['ignore', stdout, stderr] });
+    childProcessPromise.childProcess.unref();
+
+    return childProcessPromise.catch((err) => {
```

Comment:
can't this be separated into two declarations ?

```js
childProcessPromise.catch((err) => {
}
....
return childProcessPromise;
```

Reply:
But this is completely different logic then. What do you mean?
![screenshot from 2018-05-02 15-14-50](https://user-images.githubusercontent.com/1962469/39522744-dade02a0-4e1b-11e8-857e-f663f878d504.png)